keps/sig-auth/4412-projected-service-account-tokens-for-kubelet-image-credential-providers/README.md

```diff
@@ -0,0 +1,1034 @@
+    If the `ServiceAccountTokenAudience` field is configured in the kubelet's credential provider configuration, the token will be sent to the plugin.
+2. `ServiceAccountAnnotations` is a map of annotations on the KSA for which the image is being pulled. Only annotations defined in the `ServiceAccountAnnotations` field in the credential provider configuration will be passed to the plugin. If the `ServiceAccountAnnotations` field is not set in the configuration, this field will be empty.
+
+```diff
+--- a/staging/src/k8s.io/kubelet/pkg/apis/credentialprovider/types.go
++++ b/staging/src/k8s.io/kubelet/pkg/apis/credentialprovider/types.go
+@@ -32,6 +32,16 @@ type CredentialProviderRequest struct {
+        // credential provider plugin request. Plugins may optionally parse the image
+        // to extract any information required to fetch credentials.
+        Image string
++
++       // serviceAccountToken is the service account token bound to the pod for which
++       // the image is being pulled. This token is only sent to the plugin if the
++       // audience field is configured in the kubelet's credential provider configuration.
++       ServiceAccountToken string
++
++       // serviceAccountAnnotations is a map of annotations on the service account bound to the
++       // pod for which the image is being pulled. The list of annotations in the service account
++       // that need to be passed to the plugin is configured in the kubelet's credential provider
++       // configuration.
++       ServiceAccountAnnotations map[string]string
+ }
+```
+
+Example Kubernetes Service Account:
+
+```yaml
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: my-service-account
+  namespace: my-namespace
+  annotations:
+    domain.io/identity-id: "12345"
+    domain.io/identity-type: "user"
+    domain.io/annotation-that-will-not-be-passed: "value"
+```
+
+Example credential provider request:
+
+```json
+{
+  "image": "my-image",
+  "serviceAccountToken": "<service-account-token>",
+  "serviceAccountAnnotations": {
+    "domain.io/identity-id": "12345",
+    "domain.io/identity-type": "user"
+  }
+}
+```
+
+#### Caching Credentials
+
+The cache key generation will be updated to support new caching modes while preserving the existing behavior when the audience field is not set.
+
+##### Existing behavior
+
+If the `serviceAccountTokenAudience` field is not set in the provider configuration for the plugin, the key generation should fall back to the current behavior: using only the image registry URL or other existing identifiers.
+
+##### New behavior when the `serviceAccountTokenAudience` field is set
+
+Based on the `PluginCacheKeyType` definitions (Image, Registry, Global), here’s a breakdown of how cache keys would be generated:
+
+1. `GlobalPluginCacheKeyType` (Global)
+   - Global cache key type will not be allowed to use with the `serviceAccountTokenAudience` field configured for the plugin.
+2. `RegistryPluginCacheKeyType` (Registry)
```

Comment:
Which credentials will be used when I am attempting to pull an image with an SA token and the image/registry are already in the cache, but the combination of `image/registry` + SA token isn't?

Reply:
The cache is maintained per provider, so when the provider is configured with service account audience, the cache will always contain the service account as a dimension for the cache key.